File: VB/Default MVVM/FormatConditionsProvider/Converters/CellAppearanceConverter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Data
Imports System.Windows.Media
Imports System.Windows.Markup
Imports DevExpress.Xpf.Grid
Imports System.Collections.ObjectModel
Imports Default_MVVM
Imports System.ComponentModel
Imports DevExpress.Data.Filtering.Helpers

Namespace Default_MVVM
	Public Class CellAppearanceConverter
		Inherits MarkupExtension
		Implements IMultiValueConverter
		' Fields...
		Private _DefaultAppearance As Appearance

		Public Sub New()

		End Sub

		Public Property DefaultAppearance() As Appearance
			Get
				Return _DefaultAppearance
			End Get
			Set(ByVal value As Appearance)
				_DefaultAppearance = value
			End Set
		End Property

		Public Overrides Function ProvideValue(ByVal serviceProvider As IServiceProvider) As Object
			Return Me
		End Function

		Public Function Convert(ByVal values() As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements IMultiValueConverter.Convert
			Dim presenter As GridCellContentPresenter = TryCast(values(3), GridCellContentPresenter)
			Dim data As EditGridCellData = TryCast(presenter.DataContext, EditGridCellData)
			Dim view As TableView = TryCast(data.View, TableView)
			If DefaultAppearance Is Nothing AndAlso (Not presenter.IsFocusedCell) Then
				CreateDefaultAppearance(presenter)
			End If
			Dim appearance As Appearance = GetDefaultAppearance()
			Dim isFormatconditionChanged As Boolean = CBool(values(1)) OrElse CBool(values(2))
			If (Not isFormatconditionChanged) Then
				Return appearance
			End If
			Dim row As Object = values(0)
			Dim conditions As ObservableCollection(Of FormatCondition) = FormatConditionsProvider.GetFormatConditions(view)
			Dim descriptors As PropertyDescriptorCollection = TypeDescriptor.GetProperties(row)
			For Each condition As FormatCondition In conditions
				If CanApplyFormatCondition(presenter, condition, row, descriptors) Then
					Assign(appearance, condition.Appearance)
				End If
			Next condition
			Return appearance
		End Function

		Private Sub CreateDefaultAppearance(ByVal presenter As GridCellContentPresenter)
			DefaultAppearance = New Appearance()
			Assign(DefaultAppearance, presenter)
		End Sub

		Private Function GetDefaultAppearance() As Appearance
			Dim appearance As New Appearance()
			Assign(appearance, DefaultAppearance)
			Return appearance
		End Function

		Private Sub Assign(ByVal targetAppearance As Appearance, ByVal sourcePresenter As GridCellContentPresenter)
			targetAppearance.Background = sourcePresenter.Background
			targetAppearance.Foreground = sourcePresenter.Foreground
			targetAppearance.FontFamily = sourcePresenter.FontFamily
			targetAppearance.FontStyle = sourcePresenter.FontStyle
			targetAppearance.FontSize = sourcePresenter.FontSize
		End Sub

		Private Function CanApplyFormatCondition(ByVal presenter As GridCellContentPresenter, ByVal condition As FormatCondition, ByVal row As Object, ByVal descriptors As PropertyDescriptorCollection) As Boolean
			Dim evaluator As New ExpressionEvaluator(descriptors, condition.Criteria)
			Return ((String.IsNullOrEmpty(condition.FieldName) OrElse presenter.Column.FieldName.Equals(condition.FieldName))) AndAlso (Not Object.Equals(condition.Criteria, Nothing)) AndAlso evaluator.Fit(row)
		End Function

		Private Sub Assign(ByVal targetAppearance As Appearance, ByVal sourceAppearance As Appearance)
			If sourceAppearance.Background IsNot Nothing Then
				targetAppearance.Background = sourceAppearance.Background
			End If
			If sourceAppearance.Foreground IsNot Nothing Then
				targetAppearance.Foreground = sourceAppearance.Foreground
			End If
			If sourceAppearance.FontStyle IsNot Nothing Then
				targetAppearance.FontStyle = sourceAppearance.FontStyle
			End If
			If sourceAppearance.FontSize <> 0 Then
				targetAppearance.FontSize = sourceAppearance.FontSize
			End If
			If sourceAppearance.FontFamily IsNot Nothing Then
				targetAppearance.FontFamily = sourceAppearance.FontFamily
			End If
		End Sub

		Public Function ConvertBack(ByVal value As Object, ByVal targetTypes() As Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
			Throw New NotImplementedException()
		End Function
	End Class
End Namespace

File: CS/Default MVVM/FormatConditionsProvider/Converters/CellAppearanceConverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Markup;
using DevExpress.Xpf.Grid;
using System.Collections.ObjectModel;
using Default_MVVM;
using System.ComponentModel;
using DevExpress.Data.Filtering.Helpers;

namespace Default_MVVM
{
    public class CellAppearanceConverter : MarkupExtension, IMultiValueConverter
    {
        // Fields...
        private Appearance _DefaultAppearance;

        public CellAppearanceConverter()
        {

        }

        public Appearance DefaultAppearance
        {
            get { return _DefaultAppearance; }
            set { _DefaultAppearance = value; }
        }

        public override object ProvideValue(IServiceProvider serviceProvider)
        {
            return this;
        }

        public object Convert(object[] values, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            GridColumn gridColumn = values[4] as GridColumn;
            GridCellContentPresenter presenter = values[3] as GridCellContentPresenter;
            EditGridCellData data = presenter.DataContext as EditGridCellData; 
            TableView view = data.View as TableView; 
            if (DefaultAppearance == null && !presenter.IsFocusedCell)
                CreateDefaultAppearance(presenter);
            Appearance appearance = GetDefaultAppearance();
            bool isFormatconditionChanged = (bool)values[1] || (bool)values[2];
            if (!isFormatconditionChanged)
                return appearance;
            object row = values[0];
            ObservableCollection<FormatCondition> conditions = FormatConditionsProvider.GetFormatConditions(view);
            PropertyDescriptorCollection descriptors = TypeDescriptor.GetProperties(row);
            foreach (FormatCondition condition in conditions)
                if (CanApplyFormatCondition(presenter, condition, row, descriptors, gridColumn))
                    Assign(appearance, condition.Appearance);
            return appearance;
        }

        private void CreateDefaultAppearance(GridCellContentPresenter presenter)
        {
            DefaultAppearance = new Appearance();
            Assign(DefaultAppearance, presenter);
        }

        private Appearance GetDefaultAppearance()
        {
            Appearance appearance = new Appearance();
            Assign(appearance, DefaultAppearance);
            return appearance;
        }

        private void Assign(Appearance targetAppearance, GridCellContentPresenter sourcePresenter)
        {
            targetAppearance.Background = sourcePresenter.Background;
            targetAppearance.Foreground = sourcePresenter.Foreground;
            targetAppearance.FontFamily = sourcePresenter.FontFamily;
            targetAppearance.FontStyle = sourcePresenter.FontStyle;
            targetAppearance.FontSize = sourcePresenter.FontSize;
        }

        private bool CanApplyFormatCondition(GridCellContentPresenter presenter, FormatCondition condition, object row, PropertyDescriptorCollection descriptors, GridColumn column)
        {
            if(column == null) return false;
            ExpressionEvaluator evaluator = new ExpressionEvaluator(descriptors, condition.Criteria);
            return ((string.IsNullOrEmpty(condition.FieldName) || presenter.Column.FieldName.Equals(condition.FieldName)))
                && !object.Equals(condition.Criteria, null) && evaluator.Fit(row);
        }

        private void Assign(Appearance targetAppearance, Appearance sourceAppearance)
        {
            if (sourceAppearance.Background != null)
                targetAppearance.Background = sourceAppearance.Background;
            if (sourceAppearance.Foreground != null)
                targetAppearance.Foreground = sourceAppearance.Foreground;
            if (sourceAppearance.FontStyle != null)
                targetAppearance.FontStyle = sourceAppearance.FontStyle;
            if (sourceAppearance.FontSize != 0)
                targetAppearance.FontSize = sourceAppearance.FontSize;
            if (sourceAppearance.FontFamily != null)
                targetAppearance.FontFamily = sourceAppearance.FontFamily;
        }

        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, System.Globalization.CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
}